jenkins-master/jobs/scripts/workspace/runtests.py

```diff
@@ -0,0 +1,148 @@
+
+
+class Runner(object):
+
+    def __init__(self, venv_path):
+        self.venv_path = venv_path
+        self.build_path = os.path.join(here, 'build')
+
+        os.chdir(here)
+        self.activate_venv()
+
+    def activate_venv(self):
+        if not os.path.exists(self.venv_path):
+            os.chdir(os.path.join(here, 'firefox-ui-tests'))
+            subprocess.check_call(['python', 'create_venv.py',
+                                   '--with-optional-packages',
+                                   self.venv_path
+                                   ])
+            os.chdir(here)
+
+        dir = 'Scripts' if sys.platform == 'win32' else 'bin'
+        env_activate_file = os.path.join(self.venv_path, dir, 'activate_this.py')
+
+        # Activate the environment and set the VIRTUAL_ENV os variable
+        execfile(env_activate_file, dict(__file__=env_activate_file))
+        os.environ['VIRTUAL_ENV'] = self.venv_path
+
+    def download_build(self, options, args):
+        command = ['mozdownload',
+                   '--destination=%s' % self.build_path,
+                   '--type=%s' % options.build_type,
+                   '--branch=%s' % options.branch,
+                   '--build-id=%s' % options.build_id,
+                   '--locale=%s' % options.build_locale,
+                   '--platform=%s' % options.platform,
+                   ]
+        subprocess.check_call(command)
+
+    def run_tests(self, options, args):
+        # Download the build under test
+        self.download_build(options, args)
+
+        import firefox_ui_tests
+        import firefox_puppeteer
+
+        command = [
+            'firefox-ui-update' if options.type == 'update' else 'firefox-ui-tests',
+            '--installer=%s' % glob.glob(os.path.join(self.build_path, '*firefox-*'))[0],
+            '--log-xunit=report.xml',  # Enable XUnit reporting for Jenkins result analysis
+            '--log-html=report.html',  # Enable HTML reports with screenshots
+        ]
+
+        if options.type == 'update':
+            # Ensure to enable Gecko log output in the console because the file gets
+            # overwritten with the second update run
+            command.append('--gecko-log=-')
+
+            if options.update_channel:
+                command.append('--update-channel=%s' % options.update_channel)
+            if options.update_target_version:
+                command.append('--update-target-version=%s' % options.update_target_version)
+            if options.update_target_build_id:
+                command.append('--update-target-buildid=%s' % options.update_target_build_id)
+
+        elif options.type == 'functional':
+            manifests = [firefox_puppeteer.manifest, firefox_ui_tests.manifest_functional]
+            command.extend(manifests)
+
+        elif options.type == 'remote':
+            manifests = [firefox_ui_tests.manifest_remote]
+            command.extend(manifests)
+
+        print 'Execute tests: %s' % command
```

Comment:
The ´firefox-ui-tests´ command is only used for functional and remote tests. For update tests there is a ´firefox-ui-update´ command. Please be aware that exactly this command is already listed in the output.